File: library/src/main/java/me/proxer/library/api/HttpsEnforcingInterceptor.java
package me.proxer.library.api;

import me.proxer.library.util.ProxerUrls;
import okhttp3.Interceptor;
import okhttp3.Request;
import okhttp3.Response;

import java.io.IOException;

/**
 * @author Ruben Gees
 */
class HttpsEnforcingInterceptor implements Interceptor {

    @Override
    public Response intercept(final Chain chain) throws IOException {
        final Request request = chain.request();

        if (ProxerUrls.hasProxerHost(request.url())) {
            if (!request.isHttps()) {
                return chain.proceed(request.newBuilder()
                        .url(request.url().newBuilder()
                                .scheme("https")
                                .build())
                        .build());
            } else {
                return chain.proceed(request);
            }
        } else {
            throw new IllegalArgumentException("Only use ProxerLib's OkHttp instance with Proxer.Me URLs!");
        }
    }
}

File: library/src/main/kotlin/me/proxer/library/internal/interceptor/HttpsEnforcingInterceptor.kt
package me.proxer.library.internal.interceptor

import me.proxer.library.util.ProxerUrls
import okhttp3.HttpUrl
import okhttp3.Interceptor
import okhttp3.Request
import okhttp3.Response

/**
 * @author Ruben Gees
 */
internal class HttpsEnforcingInterceptor : Interceptor {

    override fun intercept(chain: Interceptor.Chain): Response {
        val request = chain.request()

        return if (ProxerUrls.hasProxerHost(request.url())) {
            if (!request.isHttps) {
                chain.proceed(request.enforceHttps())
            } else {
                chain.proceed(request)
            }
        } else {
            throw IllegalArgumentException("Only use ProxerLib's OkHttp instance with Proxer.Me URLs!")
        }
    }

    private fun Request.enforceHttps() = newBuilder()
        .url(url().enforceHttps())
        .build()

    private fun HttpUrl.enforceHttps() = newBuilder()
        .scheme("https")
        .build()
}
